volumina/tiling/cache.py

```diff
@@ -0,0 +1,295 @@
+
+    def _clean(self):
+        while len(self._caches) > self.maxsize:
+            self._evict_one()
+
+
+class TilesCache:
+    """
+    Contains the following caches, with convenience accessor functions for each.
+
+        layerCache: A cache of 'layers', i.e. for every patch a QImage or QGraphicsItem
+                    for every "image source" in the stack
+
+        tileCache: A cache of 'tiles', i.e. the blended QImage objects
+                   that were created by combining all QImage layers from layerCache for a given patch.
+                   (The QGraphicsItem layers do not contribute to the composite tiles in the tileCache.
+                   They are merely stored.)
+
+        layerCacheDirty: A cache of dirty bits for all layers in layerCache
+        layerCacheTimestamp: A cache of timestamps to track how recently each layer was needed.
+
+        tileCacheDirty: A cache of dirty bits for the composite tiles
+                        (i.e. for a given patch, if a single layer in the patch
+                        is dirty, then the tile for that patch is dirty)
+    """
+
+    def __init__(self, first_stack_id, sims, maxstacks):
+        self._lock = threading.Lock()
+        self._sims = sims
+        self._maxstacks = maxstacks
+        self._policy = CachePolicy(maxstacks)
+
+        kwargs = {"first_uid": first_stack_id, "policy": self._policy}
+
+        # [stack_id][tile_id] -> QImage or QGraphicsItem
+        self._tileCache = MultiCache(default_factory=lambda: (None, 0.0), **kwargs)
+
+        # [stack_id][tile_id] -> bool
+        self._tileCacheDirty = MultiCache(default_factory=lambda: True, **kwargs)
+
+        # [stack_id][(ims, tile_id)] -> QImage or QGraphicsItem
+        self._layerCache = MultiCache(**kwargs)
+
+        # [stack_id][(ims, tile_id)] -> bool
+        self._layerCacheDirty = MultiCache(default_factory=lambda: True, **kwargs)
+
+        # [stack_id][(ims, tile_id)] -> float
+        self._layerCacheTimestamp = MultiCache(default_factory=float, **kwargs)
+
+    @property
+    def maxstacks(self):
+        return self._maxstacks
+
+    def set_maxstacks(self, maxstacks):
+        self._policy.set_size(maxstacks)
+
+    def __enter__(self):
+        self._lock.acquire()
+        return self
+
+    def __exit__(self, *args):
+        assert self._lock.locked()
+        self._lock.release()
+
+    def __contains__(self, stack_id):
+        assert self._lock.locked(), "You must claim the _TileCache via a context manager before calling this function."
+        return stack_id in self._tileCache
+
+    def __len__(self):
+        assert self._lock.locked(), "You must claim the _TileCache via a context manager before calling this function."
+        return len(self._tileCache)
+
+    def tile(self, stack_id, tile_id):
+        assert self._lock.locked(), "You must claim the _TileCache via a context manager before calling this function."
```

Comment:
I know this is not your code, but since you are touching it anyway, we might discuss whether asserts are a good idea here. `python -O` will remove those...

Reply:
I will do a follow-up pull-request with this change. There are libraries that provide decorators for this task